[로또 - 3단계] 안석환 미션 제출합니다

seokhwan-an opened the pull request:
로또 3단계 완료해서 리뷰 요청 보냅니다.

### 미션을 진행하면서 집중한 것
- LottoNumber 객체를 만들어 Lotto의 책임 분산 (각 숫자가 1과 45 사이의 숫자인지 검증, 문자 입력이 숫자인지 검증)
- Rank에 2등 추가한 것

### 고민한 점
현재 LottoNumber는 Lotto의 각 숫자이기도 하면서 보너스 번호이기도 합니다. 현재 LottoNumber에 forBonusNumber라는 정적팩토리 메소드로 보너스 번호를 생성하며 당첨 번호에 포함되어 있지 않은지 검증이 이루지고 있습니다. 이 방식 대신에 LottoNumber를 상속받는 BonusNumber를 생성해서 앞선 검증(당첨 번호에 포함여부)을 진행하는 것은 어떻게 생각하나요? (현재 BonusNumber는 LottoNumber와 모두 같고 앞선 검증만 더 추가된 상황입니다.) 저는 이 부분에서 고민을 하다가 정적팩토리 메소드 네이밍으로도 충분이 구분이 가능할 것 같아서 LottoNumber 내부에서 보너스 번호를 생성하게 했습니다. 민주님의 의견이 궁금합니다.

seokhwan-an committed:
docs: 3단계 요구사항 추가

seokhwan-an committed:
feat: 로또 번호 객체 추가

세부사항
- 1부터 45 사이의 숫자 검증
- 문자가 숫자인지 검증
- 위의 검증 사항을 lotto -> lottoNumber로 이관

seokhwan-an committed:
feat: 보너스 볼을 입력받는 기능 추가

seokhwan-an committed:
feat: 2등 방식 추가 (당첨 번호 5개 일치 및 보너스 볼 포함)

seokhwan-an committed:
docs: 보너스 볼 필요한 요구사항 추가

seokhwan-an committed:
feat: 보너스 볼은 당첨 번호에 포함되었는 지 검증 기능 추가

seokhwan-an committed:
feat: 당첨 통계에 2등 정보 추가

seokhwan-an committed:
test: 수익률 테스트 구체화

seokhwan-an committed:
refactor: 등수 출력을 위한 분기 방식 수정

JoungMinJu commented on src/main/java/model/LottoNumber.java:
```diff
@@ -0,0 +1,49 @@
+    private final int number;
+
+    public LottoNumber(final int number) {
+        validateNumber1to45(number);
```
진짜 진짜 중요한건 아닌데 혹시 혹시 로또 번호의 min-max 조건이 바뀌게 된다면 
'validateNumber1to45'의 네이밍도 바꿔야할 것 같고
아래 "로또번호 및 보너스볼은 1과 45 사이의 숫자여야합니다."라는 string도 바꿔야할 것 같아서 유연함??이 조금 떨어지게 되는 것 같습니다!!

seokhwan-an replied:
제 생각은 유연성이 조금을 떨어지더라도 예외가 발생했을 때 정보를 명확하게 제시해주는 것이 더 좋다고 생각해서 위와 같이 예외 메세지를 제시했습니다. 이 부분에 대해서 민주님은 예외 메세지를 어떻게 처리하셨나요?

JoungMinJu replied:
lotto class에 정적변수를 놓고
```
    private boolean isNumberInRange(int number) {
        return number >= MIN_NUMBER && number <= MAX_NUMBER;
    }
```

이렇게 처리했습니다!

JoungMinJu commented on src/test/java/model/LottoNumberTest.java:
```diff
@@ -0,0 +1,34 @@
+        // then
+        assertThatThrownBy(() -> new LottoNumber(number))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("로또 번호 및 보너스 볼은 1과 45사이의 숫자이어야 합니다.");
```
요것도 에러 메세지가 변경될 때마다 테스트 코드도 수정해야할 것 같습니다..! 저는 Error 클래스를 생성하여 가져다쓰도록 구현했었는데 석환님 의견은 어떠신지 궁금합니다!

seokhwan-an replied:
현재 민주님 말에 큰 공감을 합니다. 예외 메세지가 변화하게 되면 테스트 코드도 변화를 해야하네요! 이부분은 혹시 제가 4단계에서 조금더 깊게 고민해보고 예외 메세지를 처리해보아도 될까요? 조금 생각해보는 시간을 가져봐야 할 것 같아서요!

JoungMinJu replied:
좋습니다!

JoungMinJu commented on src/main/java/controller/LottoController.java:
```diff
@@ -25,6 +25,8 @@ public void run() {
         OutputView.showLotto(transToLottosDto(lottos), lottos.getBuyLottoCount());
 
         final Lotto winningLotto = Lotto.fromStringsInput(InputView.inputWinningLotto());
+        final LottoNumber bonusNumber = LottoNumber.from(InputView.inputBonusNumber());
```
Number를 객체로 만드니까 더 로직이 직관적으로 변한 것 같아요! 배워갑니다!

seokhwan-an replied:
감사합니다!

JoungMinJu commented on src/main/java/global/Rank.java:
```diff
@@ -4,27 +4,38 @@
+
         return Arrays.stream(Rank.values())
-                .filter(r -> r.correctCnt == correct)
+                .filter(r -> r.correctCnt == correct && !r.hasBonusBall)
```
2등과 3등을 가려내려고 (2등 구분 후 비교하는) 다른 모든 rank 구분 조건에 hasBonusBall이 false여야한다는 조건을 거신게 뭔가 좀 덜 직관적인 느낌이 듭니다..! 

아예 위에서 correct가 5인 경우를 기준으로 분기처리를 하고 2등 3등으로 걸러낸 후, stream을 타면 더 낫지않을까 생각이 들었는데 그럼 depth가 깊어지겠군요 ^___^...

딱히 마땅한 해결책은 생각 안나서 문제이긴한데 :( 이와 관련해서 석환님은 어떻게 생각하시는지 궁금해서 코멘트 남겨 봅니다!!

seokhwan-an replied:
일단 filter에서 hasBonusBall이 false라는 조건을 이용한 이유는 Rank요소(first_place, second_place 등등)로 stream을 돌 때 당첨번호와 매치 개수가 5개인 경우에 위의 조건 없이 매치 개수만 이용하면 second_place가 나와 의도했던 third_place가 나타나지 않는 상황이 발생했습니다. 그래서 덜 직관적이더라도 위의 조건을 추가했는데 이 부분을 개선해볼 수 있는 방안은 조금더 고민해보겠습니다!

JoungMinJu reviewed (commented):
수고 많으셨습니다 ! :)

항상 느끼지만 커밋 기록을 진짜 깔끔하게 남겨주셔서 리뷰하기가 젤 편한 것 같습니다 배우고싶어요 🥹

코멘트 주신 내용에 대해서는 "보너스 볼 객체를 따로 만들어 검증을 분리하는 것"이 더 맞을 것 같습니다!
코드 읽으면서 "보너스 볼"이 "**추가정보**를 더 담고 있는 로또 볼", 즉 로또 볼의 하위 객체라는 생각이 들었거든요! 
석환님이 말씀해주신대로 lottoNumber를 상속받아 "보너스볼"에 특화된? 검증 로직을 추가한다면 전반적으로 로직의 가독성도 올라갈 것 같습니다! 좋은 방법인 것 같아요 

저는 전혀 생각지도 못했는데 야무진 구현 방법이 있었네요 🥹 많이 배워갑니다아아

seokhwan-an committed:
refactor: BonusNumber 객체 생성

세부사항
- LottoNumber를 상속받는 BonusNumber 객체 생성
- 당첨 번호에 보너스 볼이 포함되어 있는지 검증을 BonusNumber에서 처리

seokhwan-an commented:
민주님 제가 이번주에 예비군과 취준일정이 있었어서 리뷰를 늦게 반영했습니다. 부족했던 기간인 만큼 고민이 필요한 부분은 4단계를 진행하면서 반영해보도록 하겠습니다! 항상 꼼꼼하게 리뷰를 남겨주셔서 감사합니다!

JoungMinJu commented:
넵 일단 approve하겠습니다 :):) 의견 감사드립니다 ~!
